container block preview and inline syntax error

Jocs opened the pull request:
**Code optimization**

- [x] Optimization html preview.
- [x] Remove codeBlocks map and updateCodeBlocks method.

**Fixed**

- [x] Fix multiple math render error when change back from source code mode.
- [x] Fix #851 

**Inline syntax parse error**

- [x] Support inline syntax in multiple lines.
- [x] Support copy paste
- [x] Support import and export syntax in multiple lines.
- [x] Check updateCtrl.
- [x] Check paragraphCtrl.
- [x] Add indent code block update to updateCtrl.
- [x] Support copy soft and hard line break.

**Features**

- [x] Support multiple lines setext heading.

**existed bugs**

- [x] create blockquote and the cursor is before the char.
- [x] output markdown, pass all test.

fxha commented on src/muya/lib/eventHandler/keyboard.js:
```diff
@@ -66,10 +66,6 @@ class Keyboard {
       if (event.target.closest('[contenteditable=false]')) {
         return
       }
-      const { start, end } = selection.getCursorRange()
```
I think we still need this because the `selectionChange: expected cursor but cursor is null.` exceptions occur again. If currently is no cursor available the cursor will not be set in the next "tick"/after this function.

Jocs replied:
My fault, I thought this code was useless, so I deleted it.

fxha commented on src/muya/lib/parser/render/renderBlock/renderContainerBlock.js:
```diff
@@ -17,90 +17,77 @@ const PRE_BLOCK_HASH = {
+  if (/^h/.test(type)) {
+    if (/^h\d$/.test(type)) {
+      // TODO: This should be the best place to create and update the TOC.
+      //       Cache `block.key` and title and update only if necessary.
```
Good idea :+1:

fxha commented on src/muya/lib/parser/render/index.js:
```diff
@@ -13,6 +13,7 @@ class StateRender {
+    this.codeCache = new Map()
     this.loadImageMap = new Map()
     this.loadMathMap = new Map()
     this.mermaidCache = new Set()
```
How are these caches cleared/invalidated, may this lead to memory leaks?

Jocs replied:
This can be clear after each `render` and `partialRender`.

Jocs commented:
>open Mark Text readme and select (via double click) high_brightness, then Ctrl+C --> Error: selectionChange: expected cursor but cursor is null.

Fixed, this due to I removed the cursor null check.

>create a code clock with cpp, type <enter> and asd --> exception

How to reproduce this?

>strange inline math issue: type $$ ($<cursor>$) and press a (ok) and a again (maybe one time more) --> a lot for characters are inserted

Fixed, this due to I modify the `getTextContent` method.

>Uncaught TypeError: Cannot read property 'buildPlaceholders' of undefined when inserting a math block and type something

How to reproduce this?

fxha commented:
> How to reproduce this?

- Code Block: Write \`\`\` in preview mode, type `cpp` and press `<enter>`. Then write some text inside the code block.
- Math Block: Write empty code block and some paragraphs after the code block. Now start a math block (`$$`) and write something inside the math block.

Jocs commented:
@fxha I still can not reproduce it, 😢 Can you reproduce it everytime?

fxha commented:
Yes, here is a demo:

![mt_renderer_exceptions](https://user-images.githubusercontent.com/22716132/57180873-5e1e5100-6e8d-11e9-8338-c93e08ac2406.gif)

Jocs commented:
@fxha It seems a bug of `prismjs`, I'll look it later, now I just fallback `prismjs` to `prismjs2`.
